research-app/src/App.vue

```diff
@@ -1213,6 +1361,369 @@ export default class App extends WWTAwareComponent {
+          mode: "preloaded",
+          goto: false,
+        });
+
+        const state = this.wwtImagesetLayers[id];
+        const settings = state.settings;
+        const values: any[] = imageSetLayerSettingNames.map(setting => (settings as any)["get_" + setting]());
+        imagerySettingMessages.push({
+          event: "image_layer_set_multi",
+          id: imageset.get_name(),
+          settings: imageSetLayerSettingNames,
+          values: values,
+        });
+
+        imageryStretchMessages.push({
+          event: "image_layer_stretch",
+          id: imageset.get_name(),
+          version: 1, // Don't think that this should matter - we're only sending one per layer
+          stretch: state.scaleType,
+          vmin: state.vmin,
+          vmax: state.vmax,
+        });
+      }
+    });
+
+    const sourceMessages: selections.AddSourceMessage[] = this.sources.map(source => {
+      return {
+        type: "add_source",
+        source: this.prepareForMessaging(source),
+      };
+    });
+
+    const createAnnotationMessages: classicPywwt.CreateAnnotationMessage[] = [];
+    const annotationSettingsMessages: layers.MultiModifyAnnotationMessage[] = [];
+    for (const [id, handler] of this.annotations) {
+      const annotation = handler.annotation();
+      let shape: "circle" | "line" | "polygon" | undefined = undefined;
+      if (annotation instanceof Circle) {
+        shape = "circle";
+      } else if (annotation instanceof Poly) {
+        shape = "polygon";
+      } else if (annotation instanceof PolyLine) {
+        shape = "line"
+      }
+      if (shape == undefined) {
+        continue;
+      }
+      createAnnotationMessages.push({
+        event: "annotation_create",
+        shape: shape,
+        id: id,
+      });
+
+      let settingNames: string[] = [];
+      if (shape === "circle") {
+        settingNames = circleAnnotationSettingNames;
+      } else if (shape === "polygon") {
+        settingNames = polyAnnotationSettingNames;
+      } else {
+        settingNames = polyLineAnnotationSettingNames;
+      }
+      const values = settingNames.map(setting => (annotation as any)["get_" + setting]());
+      annotationSettingsMessages.push({
+        event: "annotation_set_multi",
+        id: id,
+        settings: settingNames,
+        values: values,
+      });
+    }
+
+    const messageStrings = [
+      coordinatesMessage,
+      backgroundMessage,
+      foregroundMessage,
+      ...loadCatalogsMessages,
+      ...catalogSettingsMessages,
+      ...loadWtmlMessages,
+      ...imageryLayerMessages,
+      ...imagerySettingMessages,
+      ...imageryStretchMessages,
+      ...sourceMessages,
+      ...createAnnotationMessages,
+      ...annotationSettingsMessages,
+    ].flatMap(s => s ? [this.encodeObjectBase64(s)] : []);
+
+    const messageString = messageStrings.join(",");
+    const outString = messageString;
+
+    const params = {
+      origin: window.location.origin,
```

Comment:
Hmmm. I think the best thing here is not the specify the `origin` query-string parameter? If someone's going to be pulling open this link later, it'll open up a freestanding version of the research app, and I think it makes the most sense for that not to allow `Window.postMessage` calls from anywhere.

Reply:
Yeah, that's a good point. I'm thinking the alternative is, if the `messages` parameter is present, to add the current origin as the allowed origin in `main.ts` (that is, on startup, rather than relying on a query parameter from when the state was serialized).